cartridge/shop/views.py

```diff
@@ -282,6 +299,15 @@ def checkout_steps(request, form_class=OrderForm, extra_context=None):
                 # and send the order receipt email.
                 order = form.save(commit=False)
                 order.setup(request)
+
+                if (not settings.SHOP_PAYMENT_STEP_ENABLED or
+                        order.total == 0):
```

Comment:
Upon closer analysis I understand what you're doing here. This view was difficult to grok to begin with. I'd have to dig in with a debugger to understand it well enough to do any further review.